### PR TITLE
Fixing a few edge cases around display name (+cleanup)

### LINKED ISSUE #13061: Strange backticks in auto-complete
The following shows strange backticks, after pressing space where the cursor is:

image

### LINKED ISSUE #13593: Exception names with backticks are incorrectly displayed
Consider the following code:
```fsharp
exception SomeError of ``thing with space``: string
```
The tooltip is displayed like this:
image
It looks strange, the backticks should be preserved in the display name.

### LINKED ISSUE #13609: Anonymous record fields with backticks are incorrectly displayed
Similar to #13593. This case is worse, because there's no color distinction between the field name and everything else:
```fsharp
type R = {| ``thing with space``: string |}
```
image

Again, backticks should be present, at least for consistency.

### LINKED ISSUE #13610: Active pattern identifiers with backticks are incorrectly displayed
Similar to #13593. Creating a separate issue so that we could have separate discussions about each case.

```fsharp
let (|``Thing with space``|_|) x = if x % 2 = 0 then Some (x/2) else None

match 4 with
| ``Thing with space`` _ -> "yes"
| _ -> "no"
```

image

For now should be displayed with backticks at least for consistency.

### PR DESCRIPTION
**Pull request by dsyme**
fixes #13061
fixes #13593
fixes #13609
fixes #13610

The cleanup is related to logical names and display names, there are also some docs added.

### CODE REVIEW

**Comment by auduchinok** [comment]
@dsyme What do you think about normalizing names for `let` bindings so they start with a lowercase letter? It'd match the core library and the most of F# code I've seen. I find the current uppercase names a bit problematic as tooling like Fantomas uses the uppercase/lowercase distinction to format applications differently to make method/constructor/union case invocations look more natural. We also try to go inline with this logic in our refactoring implementations.

I guess it should be discussed in the style guide repo instead and I'd be happy to move there, but it would be great to have these names follow the common conventions anyway.

**Comment by dsyme** [comment]
> @dsyme What do you think about normalizing names for let bindings so they start with a lowercase letter? It'd match the core library and the most of F# code I've seen. I find the current uppercase names a bit problematic as tooling like Fantomas uses the uppercase/lowercase distinction to format applications differently to make method/constructor/union case invocations look more natural. We also try to go inline with this logic in our refactoring implementations.

I can't really see us doing this.  I understand the situation but across the major Checking and CodeGen and Optimization files I've found code clearer if the major recursive and/or revealed-in-signature routines (Gen*, Optimize*, Tc*) are capitalized, while utility functions from TypedTreeOps etc. are generally camelCase (though if functionality forms a logical unit, rather than utilities, they are sometimes PascalCase)

There's no clear dividing line for this and for PrettyNaming we're sort of caught in the middle - a somewhat messy set of lowerCase names in the implementation and the PascalCase ones in the implementation.  In some ways the PascalCase indicates an intent to clean things up into a coherent whole.

File: release-notes.md [hunk 1]
```diff
@@ -35,6 +35,9 @@ These release notes track our current efforts to document changes to the F# proj
 * `SynMeasure` was extended with [SynMeasure.Paren](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synmeasure.html#Paren) case.
 * Dynamic expressions (like `x?y`) are now represented as [SynExpr.Dynamic](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synexpr.html#Dynamic) in the Untyped Syntax Tree.
 * Members with `get` and/or `set` are now represented as [SynMemberDefn.GetSetMember](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-syntax-synmemberdefn.html#GetSetMember) in the Untyped Syntax Tree.
+* `DoesIdentifierNeedBackticks` is removed, it should always be sufficient to call `NormalizeIdentifierBackticks` or else call something in `PrettyNaming`
```
Comment (dsyme) [inline release-notes.md:38]: @auduchinok @baronfel  Can you check this?  We think it is OK but would like to know if Rider or VSCode are depending on these

**Comment by dsyme** [comment]
@psfinaki So this is looking good, I'm glad to see it green

I'm surprised that no other tests failed.  There are quite a lot of places where I believe we will have fixed bugs, e.g. quick info for active pattern names that contain spaces, ```let (|``Thing with space``|_) x = ...``` and also other similar caes.

Should we do a pass together where we review the code and characterise these in test cases?  I'll go through the code now and mark up bugs I believe fixed 

thanks
don

File: src/Compiler/Service/ServiceDeclarationLists.fs [hunk 2]
```diff
@@ -196,14 +196,13 @@ module DeclarationListHelpers =
...
             let v = apref.ActivePatternVal
-            // Format the type parameters to get e.g. ('a -> 'a) rather than ('?1234 -> '?1234)
             let vTauTy = v.TauType
             // REVIEW: use _cxs here
             let (prettyTyparInst, prettyTy), _cxs = PrettyTypes.PrettifyInstAndType denv.g (item.TyparInstantiation, vTauTy)
             let remarks = OutputFullName displayFullName pubpathOfValRef fullDisplayTextOfValRefAsLayout v
             let layout =
                 wordL (tagText (FSComp.SR.typeInfoActiveRecognizer())) ^^
-                wordL (tagActivePatternCase apref.Name |> mkNav v.DefinitionRange) ^^
+                wordL (tagActivePatternCase apref.LogicalName |> mkNav v.DefinitionRange) ^^
```
Comment (dsyme) [inline src/Compiler/Service/ServiceDeclarationLists.fs:205]: This should fix quick info when hovering over `Thing with space` in 

```fsharp
let (|``Thing with space``|_) x = if x % 2 = 0 then Some (x/2) else Non

match 4 with
| ``Thing with space`` v -> "yes"
| _ -> "no"
```

Comment (dsyme) [reply src/Compiler/Service/ServiceDeclarationLists.fs:205]: Actially, no, I'm wrong.  This is using `apref.LogicalName`, it should be using `apref.DisplayName`.  Please adjust

File: src/Compiler/Service/ServiceDeclarationLists.fs [hunk 3]
```diff
@@ -224,14 +223,14 @@ module DeclarationListHelpers =
 
         | Item.RecdField rfinfo when rfinfo.TyconRef.IsFSharpException ->
             let ty, _ = PrettyTypes.PrettifyType g rfinfo.FieldType
-            let id = rfinfo.RecdField.Id
+            let id = rfinfo.DisplayName
```
Comment (dsyme) [inline src/Compiler/Service/ServiceDeclarationLists.fs:226]: This should fix quick info when hovering over `thing with space` in 

```fsharp
exception SomeError1 of message: string
exception SomeError2 of ``thing with space``: string
```

File: src/Compiler/Service/ServiceDeclarationLists.fs [hunk 4]
```diff
@@ -430,7 +429,7 @@ module DeclarationListHelpers =
 
         | Item.AnonRecdField(anon, argTys, i, _) -> 
             let argTy = argTys[i]
-            let nm = anon.SortedNames[i]
+            let nm = anon.DisplayNameByIdx i
```
Comment (dsyme) [inline src/Compiler/Service/ServiceDeclarationLists.fs:432]: This should fix quick info when hovering over `Thing with space` in 

```fsharp
type R = {| ``Thing with space``: int |}
```

**Comment by dsyme** [comment]
I notice that we didn't end up fixing the problems with `fullName*` and `FullNameOfItem` that stem from uses of `EnclosingMangledPath` (which should be renamed to `EnclosingLogicalPath`)

For example quick info for

**Comment by dsyme** [comment]
If you could add test cases for the above that would be great.  Once that's done this is ready to go :)

**Review by KevinRansom — APPROVED** [review]

File: src/Compiler/Service/ServiceDeclarationLists.fs (hunk 3 above)
Comment (psfinaki) [reply src/Compiler/Service/ServiceDeclarationLists.fs:226]: Yep, it does, here's the [issue](https://github.com/dotnet/fsharp/issues/13593), here's the [test](https://github.com/dotnet/fsharp/pull/13552/files#diff-1e6dc3c6f1f00f584bbc25dce0ef0259e996c14c91d1b58b58ace6e297325cabR495)

File: src/Compiler/Service/ServiceDeclarationLists.fs (hunk 4 above)
Comment (psfinaki) [reply src/Compiler/Service/ServiceDeclarationLists.fs:432]: Yes. The ticket is [here](https://github.com/dotnet/fsharp/issues/13609), the test is [here](https://github.com/dotnet/fsharp/pull/13552/files#diff-1e6dc3c6f1f00f584bbc25dce0ef0259e996c14c91d1b58b58ace6e297325cabR506).

File: src/Compiler/Service/ServiceDeclarationLists.fs (hunk 2 above)
Comment (psfinaki) [reply src/Compiler/Service/ServiceDeclarationLists.fs]: Yes, now this is also fixed. [Ticket](https://github.com/dotnet/fsharp/issues/13610), [test](https://github.com/dotnet/fsharp/pull/13552/files#diff-1e6dc3c6f1f00f584bbc25dce0ef0259e996c14c91d1b58b58ace6e297325cabR518).

**Comment by psfinaki** [comment]
> I notice that we didn't end up fixing the problems with `fullName*` and `FullNameOfItem` that stem from uses of `EnclosingPath` and `EnclosingMangledPath` (both of which should be renamed to `EnclosingLogicalPath`). For example quick info for hovering over `x` in
> 
> ```fsharp
> module ``Thing with space`` =
>     let x = 1
> ```
> 
> where double-ticks are not added to the full path reported. Similarly hovering over `x` here is showing the `AModule` logical name mangling:
> 
> ```fsharp
> type A = class end
> 
> module A =
>     let x = 1
> ```
> 
> Could you list this as a separate bug? I'll write some notes about how to fix it:
> 
> * Basically all the `fullName*` functions need to be duplicated out to be either `fullLogicalName*` or `fullDisplayName*` with appropriate implementations to get back the display name along the path.
> * However this is a little tricky for the *Module names which need to switch to use the `CompilationPath` instead of the `PublicPath` to pick up the `ModuleOrNamespaceKind` which indicate if module name mangling happened.
> 
> thanks

@dsyme Yep, created followup issues: #13613, #13614

File: vsintegration/tests/UnitTests/QuickInfoTests.fs [hunk 5]
```diff
@@ -428,38 +428,104 @@ module Test =
...
+let ``Automation.LetBindings.InDoBinding``() =
+    let code = """
 namespace FsTest
 
 module Test =
     do
         let fu$$nc x = ()
         ()
-""">]
-let ``Automation.LetBindings`` code =
```
Comment (psfinaki) [inline vsintegration/tests/UnitTests/QuickInfoTests.fs:459]: @KevinRansom split this theory for you :)

**Review by psfinaki — APPROVED** [review]

File: release-notes.md (hunk 1 above)
Comment (psfinaki) [reply release-notes.md:38]: So there will be followups on that PR, if we find issues we can revert some of these changes in those followups

Comment (auduchinok) [reply release-notes.md:38]: @dsyme It seems it won't affect us. I'll raise an issue if anything doesn't work as expected.

**Pull request merged**